### ISSUE TITLE
WSL2 cannot connect to localhost when the service is running on Windows

### ISSUE THREAD

**Issue by bingzhangdai** [microsoft/WSL#5211]
**Please fill out the below information:**
* Your Windows build number:  (Type `ver` at a Windows Command Prompt)
Microsoft Windows [Version 10.0.19041.264]

* What you're doing and what's happening:  (Copy&paste the full set of _specific_ command-line steps necessary to reproduce the behavior, and their output. Include screenshots if that helps demonstrate the problem.)
I am using an HTTP proxy (ShadowSocks) on windows which is listening on localhost port 1080. Then, I want the command in WSL2 to go through that proxy.
![image](https://user-images.githubusercontent.com/9895061/81916047-eb9a9200-9605-11ea-8e51-52cf8cef4a81.png)

* What's wrong / what should be happening instead:
curl: (7) Failed to connect to 127.0.0.1 port 1080: Connection refused

* Strace of the failing command, if applicable:  (If `some_command` is failing, then run `strace -o some_command.strace -f some_command some_args`, and link the contents of `some_command.strace` in a [gist](https://gist.github.com/) here).

* For WSL launch issues, please [collect detailed logs](https://github.com/Microsoft/WSL/blob/master/CONTRIBUTING.md#8-detailed-logs).

In WSL1, all things are fine. In WSL2 I could connect to the HTTP proxy through my Windows IP. I believe it is caused by WSL2 running in a separate VM.

**Comment by bingzhangdai** [comment]
To simplify the issue, I will show the telnet result, `172.21.80.1` is the Windows IP on WSL switch.
![image](https://user-images.githubusercontent.com/9895061/81917174-6d3eef80-9607-11ea-905f-a6aeb010bd39.png)

**Comment by nunix** [comment]
hi,
the issue here is that while Windows can see the `localhost` ports of WSL, the inverse in not true by default.
The ports forwarding are from WSL to Windows. If you do a `netstat -an | grep 1080` on WSL, it should return nothing.
However, when you pass it trough the Gateway (WSL switch), it will reach it.

Going forward, you can create a tunnel from Windows to WSL if you need/want to use `localhost` in your proxy (see: `netsh interface portproxy` command)

Hope this helps.
WSL Corsair

**Comment by bplasmeijer** [comment]
See also #5131 
@nunix do you see this as default behavior with the portproxy on WSL?

**Comment by bingzhangdai** [comment]
Thanks for the reply. I notice that many related issues are all about the network between the two systems. I can create a tunnel or proxy for this. But may I know if the WSL team has a specific plan or eta for the network improvement? Thanks.

**Comment by nunix** [comment]
let's wait for Build (and here I really am like you, I'm waiting on news) and see if something will done.
@bplasmeijer I remember the SSH, actually right now, `netsh` is what I use for many of my demos.

still, I do agree with you that the "reverse forwarding" from Windows to WSL2 would be a great improvement

**Comment by bplasmeijer** [comment]
> let's wait for Build (and here I really am like you, I'm waiting on news) and see if something will done.
> @bplasmeijer I remember the SSH, actually right now, `netsh` is what I use for many of my demos.
> 
> still, I do agree with you that the "reverse forwarding" from Windows to WSL2 would be a great improvement

cc: @craigloewen-msft @benhillis @sirredbeard

**Comment by gencer** [comment]
> hi,
> the issue here is that while Windows can see the `localhost` ports of WSL, the inverse in not true by default.
> The ports forwarding are from WSL to Windows. If you do a `netstat -an | grep 1080` on WSL, it should return nothing.
> However, when you pass it trough the Gateway (WSL switch), it will reach it.
> 
> Going forward, you can create a tunnel from Windows to WSL if you need/want to use `localhost` in your proxy (see: `netsh interface portproxy` command)
> 
> Hope this helps.
> WSL Corsair

@nunix can you give us an example? For example; I am trying to access `10134` port from WSL. `10134` port is on Windows assigned by Logitech Options application. I would like to access this port from inside WSL.

Is it possible with the `portproxy`? Currently my WSL IP is:  `172.27.80.1`

**Comment by eromoe** [comment]
same problem , we need to use proxy host on windows

**Comment by fatichar** [comment]
In my case, I am unable to connect to a service running on windows on ANY IP. Always the response is:
`~/tcc/network-info$ telnet  172.17.0.1 5432
Trying 172.17.0.1...
telnet: Unable to connect to remote host: Connection refused`

**Comment by eromoe** [comment]
@fatichar  turn off your firewall , it cost me serveral hours to get it out . This is difinitely another big problem

**Comment by bingzhangdai** [comment]
Actually, what I am looking for is something even more about proxy between WSL2 and Windows. I am expecting similar experiences like WSL1, where WSL and Windows are using the same address (at least users should be bothered if services on WSL and Windows are listening on different addresses) and follow the same FireWall rules.

If the network is not unified, why not use VM directly and using Samba to share the filesystem? WSL1 provide grate convenience compared with VM. 🤔

**Comment by nunix** [comment]
> > hi,
> > the issue here is that while Windows can see the `localhost` ports of WSL, the inverse in not true by default.
> > The ports forwarding are from WSL to Windows. If you do a `netstat -an | grep 1080` on WSL, it should return nothing.
> > However, when you pass it trough the Gateway (WSL switch), it will reach it.
> > Going forward, you can create a tunnel from Windows to WSL if you need/want to use `localhost` in your proxy (see: `netsh interface portproxy` command)
> > Hope this helps.
> > WSL Corsair
> 
> @nunix can you give us an example? For example; I am trying to access `10134` port from WSL. `10134` port is on Windows assigned by Logitech Options application. I would like to access this port from inside WSL.
> 
> Is it possible to do with the `portproxy`? Currently my WSL IP is: `172.27.80.1`

hi @gencer I would suppose it's possible indeed, but on `powershell` if you do a `netstat -an`, on which interface is port 10134 mapped?
if it's localhost, then WSL2 cannot reach it indeed, if it's bound to `0.0.0.0` then from WSL2 distro, try connecting to the <default gateway IP>:10134

Let me know the output of netstat and I will provide you the command

**Comment by gencer** [comment]
@nunix, Unfortunately it binds on localhost as follow:

```
  TCP    127.0.0.1:10134        0.0.0.0:0              LISTENING
  TCP    127.0.0.1:10134        127.0.0.1:50456        ESTABLISHED
  TCP    127.0.0.1:10134        127.0.0.1:50833        ESTABLISHED
  TCP    127.0.0.1:10134        127.0.0.1:56123        ESTABLISHED
```

However, we make it work on SSH/WSL by making vscode extension runing on UI side. This makes us to access localhost port and data.

**Comment by djfos** [comment]
you can access Windows port from WSL2 using your local IP like 192.168.31.66 instead of localhost, though is not elegant

**Comment by toryano0820** [comment]
I also have experienced this issue when accessing my Windows' Redis instance from my project running in WSL2. End up mapping Nameserver to localhost in `/etc/hosts`.

You can find Nameserver from `/etc/resolv.conf`, it looks like: `nameserver 172.18.144.1`
Then you can add line in `/etc/hosts` like: `172.18.144.1 localhost`

Created a shell script to make life simpler: https://gist.github.com/toryano0820/6ee3bff2474cdf13e70d972da710996a

Not sure yet if there's a downside with this method. But it works!

**Comment by bsplosion** [comment]
This is a complete duplicate of #4619, and all details there are applicable here.  For those not wanting to scroll through everything there, there doesn't seem to be any indication that WSL2 will go back to handling networking in a similar manner to WSL1.  Your two options are 1: Put together some hacks and maybe it'll work, or 2: Revert to WSL1 and hope WSL2 maybe someday changes its approach back to that of WSL1.

**Comment by itsakt** [comment]
I was having similar issues. For me it's a hit or miss scenario. Sometimes it works sometimes it doesn't. I started webpack dev server (0.0.0.0:500) and accessed it through 127.0.0.1:500, usually browser will wait for bundles to load but today it loaded html page then just disconnected. Reloading no longer loads the page.

I had to disable my vpn before starting local server. After that turning on VPN seems to have no impact on server. I tried shutting down wsl instance (VPN still connected) and starting it again and everything seems to work fine. 🤔

**Comment by hasan-hasanov** [comment]
I can confirm that I can connect to the WSL2's localhost using VPN but are unable to do so when I am disconnected.

**Comment by troll-os** [comment]
I personally ended up running containers with the WSL2 integration for services that can run on Docker (Mongo for example)

**Comment by medmin** [comment]
This is so frustrating

**Comment by pgwilliams** [comment]
When I upgraded my distro I started having all sorts of DNS problems (even with McAfee firewall turned off) and then today hit this issue of not being able to connect to a port of a process running in Windows from my Ubuntu command line.

I've reverted the distro to WSL 1:    wsl.exe --set-version Ubuntu 1    and things seem to be back to normal now.

**Comment by DachuanZhao** [comment]
> @fatichar turn off your firewall , it cost me serveral hours to get it out . This is difinitely another big problem

It works . But I want to know why . I have added advanced rule in firewall , But It doesn't work .

**Comment by jorgeorpinel** [comment]
So you can't run SSH or any daemons (https://github.com/microsoft/WSL/issues/994#issuecomment-242918148) and you can't connect to servers run as Windows services (SSH e.g. #3173). 😢

**Comment by spyro2000** [comment]
This is just a nightmare. Was expecting to work this just out of the box. Can't believe it doesn't.

**Comment by MarcLongJoyride** [comment]
How is this still not resolved after all these people complaining of this issue for over a year? This is such a problem, my Linux coworkers keep telling me to drop WSL and just dual boot.

**Comment by bingzhangdai** [comment]
I switched to WSL1. WSL1 even performs better under Windows file system. If you choose to use Linux, you still cannot visit Windows ports through localhost.

**Comment by chupzzz** [comment]
**May 2021 - we're still here!** Unexpected behavior of WSL2 networking after each restart of my PC. Everytime I have NEW network issues.

That's terrible to pretend WSL2 is stable and recommend it as default! **Without stable networking all the fancy features of WSL2 is totally useless!!!** Microsoft, please put efforts on this. First things - first!

**Comment by samliddicott** [comment]
I finally got a custom wsl2 kernel with NBD support only to find out I can't connect to the NBD server running under windows.

(I'm trying to mount my ext4 partition in wsl2)

**Comment by danivalls** [comment]
Hey! I was having the same issue and I finally fixed it.

Opening a PowerShell and `wsl --shutdown` worked for me like a charm! (Make sure to have closed all your wsl terminals first)

Hope it helps!

**Comment by rifton007** [comment]
> Hey! J'avais le même problème et je l'ai finalement résolu.
> 
> `wsl --shutdown`sur le PowerShell a fonctionné pour moi comme un charme ! (Assurez-vous d'avoir fermé tous vos terminaux wsl au préalable)
> 
> J'espère que cela aide!

That works, but you need  a shutdown/restart WSL2 after each launching Windows ?
Someone knows why ?

**Comment by J4CKVVH173** [comment]
Microsoft never changes

**Comment by MarcLongJoyride** [comment]
What's bizarre is that my desktop does not have this issue but my laptop has this issue at least once a week. 

It's incredibly frustrating to get everything setup and run npm start just to realize you have to shut the entire wsl environment down and start all over again.

**Comment by MarcLongoria** [comment]
What's bizarre is that my desktop does not have this issue but my laptop has this issue at least once a week.

It's incredibly frustrating to get everything setup and run npm start just to realize you have to shut the entire wsl environment down and start all over again.

**Comment by samwightt** [comment]
Is there a reason why this hasn't received more attention? This is really, really annoying.

**Comment by 2pl** [comment]
This behavior is insane and has one killer side effect: random and unpredictable service ports conflicts.

Let's assume you have a windows process listening on 127.0.0.1 and for example port 9000

Port is visible from Windows 
```
❯❯❯ NETSTAT.EXE -an | grep LISTEN | grep 9000
  TCP    0.0.0.0:9000           0.0.0.0:0              LISTENING
  TCP    127.0.0.1:9000         0.0.0.0:0              LISTENING
❯❯❯

```
but from WSL2 you are **NOT** seeing it:

```
❯❯❯ netstat -an | grep LISTEN | grep 9000
❯❯❯

```
Because you are not seeing it, then a wsl2 process can legitimately open the same port for listen, thinking that no other process is using it, like for example (note the absence of reuseaddr and reuseport)
```
❯❯❯ socat TCP4-LISTEN:9000,bind=127.0.0.1,fork exec:'/usr/bin/echo'
```
But this is now visible from both WSL2 and windows ! Two listeners on 127.0.0.1:9000
```
❯❯❯ NETSTAT.EXE -an | grep LISTEN | grep 9000
  TCP    0.0.0.0:9000           0.0.0.0:0              LISTENING
  TCP    127.0.0.1:9000         0.0.0.0:0              LISTENING
  TCP    127.0.0.1:9000         0.0.0.0:0              LISTENING
❯❯❯

```
**And the WSL2 port can be randomly hit instead of the Windows one causing unpredictable behavior.** 

From my experience the Windows port seems to be 'winning' most of the time, but if the port is under heavy use it sometimes happen that the connection is accepted by the WSL2 listener.

**Comment by bingzhangdai** [comment]
> From my experience the Windows port seems to be 'winning' most of the time, but if the port is under heavy use it sometimes happen that the connection is accepted by the WSL2 listener.

Exactly. Even Windows is not listening on port 9000, connecting to port 9000 on Windows will not always hit port 9000 in WSL2. It sometimes says connection refused. This situation happens where sshd is running in WSL2, and I want to `ssh localhost` in Windows.

**Comment by AlexZablotsky450501** [comment]
I also saw similar behavior. We are running Docker Desktop with WSL2 feature enabled. We created a container with 443 port exposed. And for some reason after 2 days of stable work the port become inaccessible. One more strange thing for me is that this port is inaccessible when I try to reach it from another host from the same network (https://host) or even from the same host via its IP (https://192.168.x.x), but it is accessible when I using localhost (https://localhost). Does anyone has any ideas what is happening?)

**Comment by bingzhangdai** [comment]
@AlexZablotsky450501 Probably, when Windows is forwarding the listening port from WSL2 to Windows, the listening address is changed to localhost.

Runthis command in WSL2: `socat TCP4-LISTEN:2333,bind=0.0.0.0,fork exec:'/usr/bin/echo'`. Then, I opened another terminal and run `lsof -i | awk "NR==1 || /LISTEN/"`, I get,
```bash
COMMAND PID  USER   FD   TYPE DEVICE SIZE/OFF NODE NAME
socat   277 david    5u  IPv4  17568      0t0  TCP *:2333 (LISTEN)
```
It looks fine. But running `NETSTAT.EXE -an | grep LISTEN | grep 2333`, I only get
```bash
  TCP    127.0.0.1:2333         0.0.0.0:0              LISTENING
```
It seems that on Windows, it is listening on localhost:2333 not 0.0.0.0:2333. This is why you cannot connect from another host. I also double checked that it indeed cannot be connected from another host.

**Comment by PengyuApp** [comment]
It happens from randomly. Usually `wsl --shutdown` and restart fix the issue. Very annoying, hope there can be a patch for this issue.

**Comment by thomas-lm** [comment]
It may be useful for you :
I had the same issue when i use VPN on windows to connect to Postgresql (port 5432) server using podman in wsl2/ubuntu. This issue not append with VPN turned off.

It failed on localhost or local ip connection.
BUT 
running NETSTAT.EXE -an | grep LISTEN | grep 5432 show me that listening port is on ::1 (ipv6) and not 127.0.0.1. 
Then this issue not append anymore when i used ::1 instead of localhost

**Comment by Nayrito** [comment]
Es muy molesto , pero al reiniciar wsl2 nuevamente puedo acceder a localhost : https://snippets.khromov.se/how-to-restart-wsl2-windows-subsystem-linux-on-windows-10/

**Comment by lcaparros** [comment]
I have checked I have exactly the same issue when connecting through the VPN. I have tried restarting wsl2 and changing the IP address to ::1 as commented above, but nothing works. I will write back if I find a solution.

**Comment by samon11** [comment]
Workaround for those looking to just run a bash command against your service such as `curl` you can access localhost through git bash

**Comment by RaymondFallon** [comment]
FYI for folks following this thread: after the [latest Windows update](https://support.microsoft.com/en-us/topic/september-14-2021-kb5005565-os-builds-19041-1237-19042-1237-and-19043-1237-292cf8ed-f97b-4cd8-9883-32b71e3e6b44), this was no longer broken for me. I don't see this issue mentioned anywhere in the update notes, but it may be worth your time to see if this is now fixed for you, as well.

**Comment by jhgorse** [comment]
Can confirm that WSL2 does NOT connect to Windows 10 services via 'localhost' hostname or IP addresses for the following version:

```
Microsoft Windows [Version 10.0.19042.1237]
(c) Microsoft Corporation. All rights reserved.

C:\Users\Joe Gorse>ver

Microsoft Windows [Version 10.0.19042.1237]
```

I can confirm that migrating to WSL1 resolves the issue entirely. For a comparison of features, see https://docs.microsoft.com/en-us/windows/wsl/compare-versions

Notably, WSL1 outperforms WSL2 for Windows File access performance. While WSL2 has the "full Linux kernel", it lacks serial device support that works for WSL1. See the link for full details.

Cheers,
Joe

**Comment by hfingler** [comment]
None of the tricks shown here worked for me and I can't switch the listening IP since it's not my code.
Rolling back to WSL1 worked, though. I just followed @jhgorse 's comment and ran on a powershell:

`wsl.exe --set-version Ubuntu-20.04 1`

**Comment by cip8** [comment]
This happens *very often* on my end too (Windows 11 w/ WSL2). Running `wsl --shutdown` helps, but only for 15 minutes, until I get a new `ERR_CONNECTION_REFUSED` for localhost.

Any fix yet, other than reverting to `WSL1`?

**Comment by YuraNick** [comment]
My laptop has ip 192.168.0.6 in local net. And access the internet from a wsl2. So, I can use ip 192.168.0.6 as an external ip address for a laptop on my LAN. I have web server on 8080 port. And into wsl2 I can run:
`curl http://192.168.0.6:8080`
It works!

Difficulties:
- I have to to open port (8080 for example and other used) in Windows 10. This reduces security.
- I have to have a static ip for my laptop on the network or I have to change the dynamic ip and keep this in mind.

Connect from wsl2 to VirtualBox use ssh may be with port mapping use in VirtualBox. For example port mapping tcp 192.168.100.6:2222 to 22. Don`t forget open 2222 port on Windows. Then in wsl2 run:
`ssh <user>@192.168.0.6 -p 2222`

Inconvenient but it works! 
I would like a simple solution to appear.

**Comment by SongShaopu1998** [comment]
1. In **Powershell**, use `ipconfig` command to find the **IPV4** address in  **vEthernet (WSL)**.
2. Add the normal proxy setting command in **~/.bashrc** (e.g. export http_proxy="http://172.65.25.3:1080") for http & https using the address you found in the above step.
3. `source ~/.bashrc`
4. Close the current **wsl** terminal and open it again. `curl www.google.com` should work.
> In the ssr software, use **Socks5(support UDP)** & **Allow connections from LAN**.

**Comment by kiot389** [comment]
What about setting up SSH Tunneling for this? _i tried to solve this last year. i had to recompile the wsl distro, because i read it pulls down network settings for apt functionality_

**Comment by rayfoss** [comment]
No one here seems to have brought this up... so I will:

In ChromeOS's version of WSL2 called Crostini, all listening ports are automatically forwarded to the host machine and accessible at localhost:1234 or example.localhost:1234. The Android VM doesn't support that, but it has a stable IP address.

P.S.
If I don't figure this out in the next 5 minutes I'm shifting my resources to dual booting Manjaro Gnome... for context I've used ChromeOS for development primarily on my Laptop and Ubuntu on desktop. My new laptop is too fancy for stock Manjaro linux... but I found the Kernel driver that I needed to use and OLED damage will be someone else's problem. The value proposition of Windows has been so negative, I might just erase it and dual boot SteamOS/Manjaro Gnome.

**Comment by fadebowaley** [comment]
> I can confirm that migrating to WSL1 resolves the issue entirely. For a comparison of features, see https://docs.microsoft.com/en-us/windows/wsl/compare-versions
> 
> Notably, WSL1 outperforms WSL2 for Windows File access performance. While WSL2 has the "full Linux kernel", it lacks serial device support that works for WSL1. See the link for full details.
> 
> The conversion command where "Ubuntu-20.04" may be any from the list wsl.exe -l -v:
> 
> wsl.exe --set-version Ubuntu-20.04 1

**Comment by manulep** [comment]
Not wanting to switch to WSL1, I put my server in a docker container. It works for me.

**Comment by pkaushal-mdsol** [comment]
The application running on WSL2 was unable to access API running on windows host at `localhost:5001` 

This worked for me.
1) Run ipconfig on windows host
2) Get the IPV4 Address from `Ethernet adapter vEthernet (WSL)`  - Eg - 172.22.176.1
3) Replace `localhost:5001`  with `172.22.176.1:5001`

**Comment by githubkuyaya** [comment]
Why is this taking so long, this can't be that hard to fix, can it? It's just annoying having to deal with stuff like this.

**Comment by gpotter2** [comment]
On my end this happened because I was using the WSL Insider Preview from the store. Uninstalling it fixed the issue.

**Comment by helgatheviking** [comment]
@gpotter2 I just installed that here as well. Did uninstalling it remove any of the files you had saved in your linux subsystem?

**Comment by gpotter2** [comment]
@helgatheviking It did not !

**Comment by kiot389** [comment]
I found an old article about localhost and bind 0.0.0.0. I'm posting it in case it can lead anyone who's troubleshooting to something new. https://www.bleepingcomputer.com/news/security/wsl2-now-supports-localhost-connections-from-windows-10-apps/

**Comment by githubkuyaya** [comment]
> I found an old article about localhost and bind 0.0.0.0. I'm posting it in case it can lead anyone who's troubleshooting to something new. https://www.bleepingcomputer.com/news/security/wsl2-now-supports-localhost-connections-from-windows-10-apps/

Thanks for the input. However, being able to access a listener/server started in WSL is not the problem, it's the other way around. You can't access servers started in Windows with WSL.
This here is what the issue is about: 
![image](https://user-images.githubusercontent.com/46228349/185872121-328a1216-87d0-4e22-a203-8c563c4f5580.png)

**Comment by doctorchuks** [comment]
This Microsoft page helped solve the problem on my end

https://docs.microsoft.com/en-us/windows/wsl/networking

Check in the "Accessing Windows networking apps from Linux (host IP)" section

Works pretty fine👌

**Comment by githubkuyaya** [comment]
> This Microsoft page helped solve the problem on my end
> 
> https://docs.microsoft.com/en-us/windows/wsl/networking
> 
> Check in the "Accessing Windows networking apps from Linux (host IP)" section
> 
> Works pretty fine👌

Yes, this was already mentioned in this thread a few times (e.g. https://github.com/microsoft/WSL/issues/5211#issuecomment-706652801)
Not a fix for the problem though, the issue is that one should be able to access it through `localhost`, not the IP.

**Comment by bodinsamuel** [comment]
Did anyone got around "connection refused" issue even when using the `nameserver` ip?

I tried: disabled firewall, fast startup, allowed `localhostForwarding`, rebooted everything... 
Still can't reach windows from WSL2 :(

```sh
cat /etc/resolv.conf
> nameserver 172.24.176.1

curl -XGET http://172.24.176.1:4040/inspect/http -I -v
*   Trying 172.24.176.1:4040...
* TCP_NODELAY set
* connect to 172.24.176.1 port 4040 failed: Connection refused
* Failed to connect to 172.24.176.1 port 4040: Connection refused
* Closing connection 0
curl: (7) Failed to connect to 172.24.176.1 port 4040: Connection refused
```

**Comment by kiot389** [comment]
@bodinsamuel does your windows host use a proxy lan setting or .pac script? if so, any recent changes?

**Comment by JuryA** [comment]
I also need some kind of localhost proxy from WSL2 to Windows host - in case I want to execute Windows binaries (that´s why WSL to Windows Interop exists), I cannot use IP Address. For example, if I Packer starts VirtualBox VM with port forwarding, it works until the moment when starts waiting for localhost:port. Boom! It ends in a never-ending loop...

**Comment by lalilaloe** [comment]
I basically avoided this issue by using WSL1 for the past years, but I found a solution that works for me; If you don't encounter this issue often just use $(hostname).local 😒

TLDR; This changes `localhost` to point to your Windows IP, `127.0.0.1` is still available for wsl. In windows you can access WSL via `localhost` as usual, because windows listens for exposed ports on WSL. Exposes your set windows port(s) via `portproxies`.

1. Add this to your `/etc/wsl.conf`, 

```bash
[boot]
command="sed -i \"s/127.0.0.1$(printf '\t')localhost/$(tail -1 /etc/resolv.conf | cut -d' ' -f2)$(printf '\t')localhost/g\" /etc/hosts 2>&1"
```
This command replaces the default `localhost` in `/etc/hosts` to point to your Windows IP, `127.0.0.1` is still available for wsl. This might require setting the host in your code to `127.0.0.1` for example in `vite` add `server: { host: "127.0.0.1" }` . Services running on Windows might need to be bind to 0.0.0.0 or it's local ip for a service to be accessible in WSL.

 2. a. And use this powershell script to configure portproxies [fix-wsl2-port-forwading-to-windows-localhost](https://rpfilomeno.me/powershell/fix-wsl2-port-forwading-to-windows-localhost/). The firewall rules, might not be necessary but can be helpful if you encounter issues.

       b. While the above script also helps you set the firewall rules. You can also execute the following command to set the portproxies manually in (admin) Powershell. In this example ports 8080,27017. (credits above)

```batch
$addr='0.0.0.0'; $ports=@(8080,27017); # set ports according to your use case, what ports from windows should be exposed.
$remoteport = (wsl hostname -I).Trim(); for( $i = 0; $i -lt $ports.length; $i++ ){ $port = $ports[$i]; echo "added portproxy ${addr}:${port} to ${remoteport}:${port}"; iex "netsh interface portproxy delete v4tov4 listenport=$port listenaddress=$addr"; iex "netsh interface portproxy add    v4tov4 listenport=$port listenaddress=$addr connectport=$port connectaddress=$remoteport"; }
```
You can reset these proxies anytime using `netsh interface portproxy reset`

This for me solves many of the main issues why I didn't switch, it's not the same as using WSL1. But does the job since you don't have to change `localhost` host configurations for any of your projects. Which is [frustrating](https://github.com/microsoft/WSL/issues/5211#issuecomment-751945803) when using WSL2. Bonus you can enjoy the latest features such as `WSLg` and better performance 😀.

**Comment by korneel** [comment]
As a workaround, I added the following to my `.zshenv` file to set the IP address of the Windows host to an environment variable:

```bash
if [[ -n "$IS_WSL" || -n "$WSL_DISTRO_NAME" ]]; then
  if [[ -z "$WINDOWS_HOST_IP_ADDRESS" ]]; then
    export WINDOWS_HOST_IP_ADDRESS="$(powershell.exe -Command '(Get-NetIPConfiguration | Where-Object { $_.IPv4DefaultGateway -ne $null -and $_.NetAdapter.Status -ne "Disconnected"}).IPv4Address.IPAddress' | tr -d '\r')"
  fi
fi
```

This way I can use the `WINDOWS_HOST_IP_ADDRESS` environment variable to proxy requests from WSL2 to the Windows host. For example, I run the Vite development server and proxy API requests to a .NET web application running on the Windows host.

**Comment by osmial** [comment]
Had the same issue, mentioned workarounds relying only on IPs from /etc/resolv.conf or /etc/hosts are close but not enough, at least for me. Anyone who have this problem should try finding correct IP, for me it was one of those printed by `nslookup $(hostname)`.

**Comment by mfwhatever** [comment]
Posting my solution to a similar problem.

I was trying to find a way to access RStudio Server (running on WSL2 Ubuntu) from another computer on my local network. RStudio Server is accessible on the Windows Host at localhost:8787. I needed to do both of these:

1) Create an [inbound rule on Windows Defender Firewall](https://learn.microsoft.com/en-us/windows/security/operating-system-security/network-security/windows-firewall/create-an-inbound-port-rule) for port 8787

2) Create a forwarding rule for incoming connections to port 8787 to be directed to localhost
Adapted from [here](https://learn.microsoft.com/en-us/windows/wsl/networking#accessing-a-wsl-2-distribution-from-your-local-area-network-lan). The guide suggests forwarding to the local windows address for the WSL2 VM (`wsl hostname -i`), which may be the right thing for other people, but in my case RStudio Server is already forwarded to localhost:8787, so that didn't work. I used 127.0.0.1 instead and now it works.

**Comment by pwang2** [comment]
This is a frustrating topic.  Exposing all the interfaces from windows via 0.0.0.0 is INSECURE.  it would be great and makes most sense to just keep it behave the same as a loopback interface in WSL..

**Comment by ludobegins** [comment]
That moment when you scroll the whole thread since 2020 hoping at this point someone would have a good fix but nope 😂😕

**Comment by MarcLongoria** [comment]
> That moment when you scroll the whole thread since 2020 hoping at this point someone would have a good fix but nope 😂😕

No joke. I've been following this thread since it was opened, hoping someone far more intelligent than me would have solved this by now. Yet here we are lol

**Comment by obedm503** [comment]
This was fixed in [#4619](https://github.com/microsoft/WSL/issues/4619#issuecomment-1778054020)

**Comment by dons20** [comment]
@obedm503 Thanks for pointing it out. Would have never known it was actually being addressed all this time.

Had to take a few steps to get it working on my local environment. Just thought I'd share. The changes are on a pre-release of WSL so it won't be available through regular update paths.

1. Update Windows to at least the October 10th build. See more info here: https://github.com/microsoft/WSL/issues/4619#issuecomment-1778054020
2. Update your WSL to the latest pre-release by running this in CMD or Powershell `wsl --update; wsl --update --pre-release`. You can view the changes of the pre-release versions here: https://github.com/microsoft/WSL/releases
3. Update (or create) your `.wslconfig` file which is located at `%USERPROFILE%` in explorer. Add the following:
    ```
    [wsl2]
    networkingMode=mirrored # add this line if file already exists
    ```
4. (Optional) Shutdown and restart your WSL (after a few seconds) so that the config changes apply.

Since getting it enabled, I haven't had a single issue with my localhost being inaccessible. I could safely enable IPV6 again too.

Side note, pre-release v2.0.5 https://github.com/microsoft/WSL/releases/tag/2.0.5 moved the `networkingMode` setting from `[experimental]` to `[wsl2]`.

**Comment by seri0zha** [comment]
@dons20 thank you so much, it works fine for me. I have spent a lot of time trying to resolve this problem

**Comment by MarcLongoria** [comment]
Awesome to hear! The release mentions it's a window 11 fix. Anyone on Windows 10 to confirm it works for them as well?

**Comment by Jacksonmills** [comment]
> @obedm503 Thanks for pointing it out. Would have never known it was actually being addressed all this time.
> 
> Had to take a few steps to get it working on my local environment. Just thought I'd share. The changes are on a pre-release of WSL so it won't be available through regular update paths.
> 
>     1. Update Windows to at least the October 10th build. See more info here: [[WSL 2] WSL 2 cannot access windows service via localhost:port #4619 (comment)](https://github.com/microsoft/WSL/issues/4619#issuecomment-1778054020)
> 
>     2. Update your WSL to the latest pre-release by running this in CMD or Powershell `wsl --update; wsl --update --pre-release`. You can view the changes of the pre-release versions here: https://github.com/microsoft/WSL/releases
> 
>     3. Update (or create) your `.wslconfig` file which is located at `%USERPROFILE%` in explorer. Add the following:
>        ```
>        [wsl2]
>        networkingMode=mirrored # add this line if file already exists
>        ```
> 
>     4. (Optional) Shutdown and restart your WSL (after a few seconds) so that the config changes apply.
> 
> 
> Since getting it enabled, I haven't had a single issue with my localhost being inaccessible. I could safely enable IPV6 again too.
> 
> Side note, pre-release v2.0.5 https://github.com/microsoft/WSL/releases/tag/2.0.5 moved the `networkingMode` setting from `[experimental]` to `[wsl2]`.

This worked for me!

**Comment by teapot2** [comment]
@dons20 

This worked for me too on Windows 11. Thank you!

**Comment by derharry** [comment]
Wow! I just did the steps and finally I can connect to local host services again while in Public LAN or Wifi. 
I always had to set the public LAN/Wifi to private (but that did not feed ok exposing the services to everyone in the LAN), I tried playing around with the Firewall, or I switched to private-hotspot (when I had enough connectivity) . 

Thanks!! :-)
Finally I can set most LANs/Wifis to public again :-)

**Comment by derharry** [comment]
@dons20  Thanks! Works for me on W11. 

This issue, WSL cannot connect to remote localhost, on my side had to do when the connected WiFi or LAN was set to "public network". On "private network" everything was fine. 
With this fix  WSL also connects to localhost while in public network.  :-)

**Comment by Vividious** [comment]
@dons20 cheers mate!
For real, over a year of struggle

**Comment by mamlzy** [comment]
> **Update since [Major Version 2.0.9](https://github.com/microsoft/WSL/releases/tag/2.0.9https://github.com/microsoft/WSL/releases/tag/2.0.9)**
> 
> For step 2, you can just do `wsl --update` only, no need to do the pre-release anymore.
> 
> Original
> 
> @obedm503 Thanks for pointing it out. Would have never known it was actually being addressed all this time.
> 
> Had to take a few steps to get it working on my local environment. Just thought I'd share. The changes are on a pre-release of WSL so it won't be available through regular update paths.
> 
> 1. Update Windows to at least the October 10th build. See more info here: [[WSL 2] WSL 2 cannot access windows service via localhost:port #4619 (comment)](https://github.com/microsoft/WSL/issues/4619#issuecomment-1778054020)
> 2. Update your WSL to the latest pre-release by running this in CMD or Powershell `wsl --update; wsl --update --pre-release`. You can view the changes of the pre-release versions here: https://github.com/microsoft/WSL/releases
> 3. Update (or create) your `.wslconfig` file which is located at `%USERPROFILE%` in explorer. Add the following:
>    ```
>    [wsl2]
>    networkingMode=mirrored # add this line if file already exists
>    ```
> 4. (Optional) Shutdown and restart your WSL (after a few seconds) so that the config changes apply.
> 
> Since getting it enabled, I haven't had a single issue with my localhost being inaccessible. I could safely enable IPV6 again too.
> 
> Side note, pre-release v2.0.5 https://github.com/microsoft/WSL/releases/tag/2.0.5 moved the `networkingMode` setting from `[experimental]` to `[wsl2]`.

Works! thank you so much!

**Comment by pmartincic** [comment]
As mentioned by this comment: https://github.com/microsoft/WSL/issues/5211#issuecomment-1783883702 fixed by mirrored mode.

**Comment by misternian** [comment]
@dons20 save my time. Thank you very much!
